### PR TITLE
NutritionScore Calculation Details HTML Template

### PR DESCRIPTION
**Pull request by areeshatariq**
**Description:**
HTML Template has been created for Nutrition Score Calculation Details using Template::Toolkit
The template file is used for the presentation purpose only, while all the logic has been kept in /lib/ProductOpener/Display.pm

**Related issues and discussion:** #2416

### CODE REVIEW

**Comment by stephanegigandet** [comment]
Note for reviewers: @areeshatariq is working on using a template engine for the HTML code generated by Product Opener. Copying here Areeesha's message in #productopener on Slack:

"I've opened a Pull request on HTML Template of Nutrition Score Calculation Details. Started with Nutrition Score as it was a more independent module. We used Template::Toolkit for this. What do you think of the approach and how the template looks?"

File: lib/ProductOpener/Display.pm [hunk 1]
```diff
@@ -8634,79 +8635,101 @@ the rounded value according to the Nutri-Score rules, and the corresponding poin
...
+				maximum => scalar(@{$points_thresholds{$nutrient_threshold_id}}),
+				value => $nutriscore_data_ref->{$nutrient},
+				rounded => $nutriscore_data_ref->{$nutrient . "_value"},
+			};
+		 }
+		 
+		 push @{$template_data_ref->{points_groups}}, $points_group_ref;
+	 }
+	# Nutrition Score Calculation Template
+		my $config = {
```
Comment (VaiTon) [inline lib/ProductOpener/Display.pm:8720]:
```suggestion
	my $config = {
```
Also, aren't these config we can use for every template? Can't we put them outside of the sub?

File: templates/nutrition_score.tt [hunk 2]
```diff
@@ -0,0 +1,41 @@
+
```
Comment (VaiTon) [inline templates/nutrition_score.tt:1]:
```suggestion
```

File: templates/nutrition_score.tt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  [% END %]
+
+    <p>[% nutriscore_protein_info %] </p>
+    <p>
+      <strong> [% lang('nutriscore_score') %][% lang('sep') %]: [% score %] </strong> ( 
+      [% negative_points %] - [% positive_points %])
+    </p>
+    <p><strong> [% lang('nutriscore_grade') %][% lang('sep') %]: [% grade %] </strong></p>
+
+</div>
```
Comment (VaiTon) [inline templates/nutrition_score.tt:41]:
```suggestion
</div>

```

File: templates/nutrition_score.tt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  [% FOREACH group IN points_groups %]
+
+    <p> 
+      <strong> [% lang("nutriscore_" _ group.type _ "_points") %][% lang('sep') %]: [% group.points %]  </strong> 
+    </p>
+
+    <ul>
+      [% FOREACH nutrient IN group.nutrients %]
+        <li>
+          <strong> [% lang("nutriscore_points_for_" _ nutrient.id) %][% lang('sep') %]: [% nutrient.points %]&nbsp</strong>
```
Comment (VaiTon) [inline templates/nutrition_score.tt:25]: `[% lang("nutriscore_points_for_" _ nutrient.id) %]`

What's the `_` before `nutrient.id` for?

Comment (areeshatariq) [reply templates/nutrition_score.tt:25]: @VaiTon it's for concatenation

Comment (Ban3) [reply templates/nutrition_score.tt:25]: @VaiTon The underscore operator is explained at least on this part of the manual: http://template-toolkit.org/docs/manual/Directives.html#section_SET

Maybe the double quoted string interpolation would be clearer.

Comment (VaiTon) [reply templates/nutrition_score.tt:25]: > In Perl 5, the . is used for string concatenation, but in Perl 6, as in the Template Toolkit, the . will be used as the method calling operator and ' _ ' will be used for string concatenation.

I was used to the `.` for concatenation in Perl.

> Maybe the double quoted string interpolation would be clearer.
I agree. The others? @openfoodfacts/open-food-facts-contributors

**Comment by CharlesNepote** [comment]
That's a VERY good start, thank you @areeshatariq!

The Template::Toolkit is not as usable as I would like, but it's a good enhancement.

PROs:
* I'm able to read the template and work on it easily (for example, I would like to add CSS classes or ids to some paragraphs)
* the template is/can be prettified for better readability and ease of use (indentation etc.)

CONs:
* Github doesn't provide syntax highlighting for the template

Questions:
* is it possible to insert comments in the template that won't be rendered in the HTML: I think it's very important to let back-end devs document some things
* is it possible to validate HTML code during the pull request workflow
* shouldn't we avoid conditionals IF-THEN-ELSE as much as we can: e.g.
```
  [% IF is_beverage %]
    <p> [% lang('nutriscore_is_beverage') %] </p>
  [% ELSE %]
    <p> [% lang('nutriscore_is_not_beverage') %] </p>
  [% END %]
```
shouldn't we compute a `nutriscore_beverageness` in the back-end?

File: templates/nutrition_score.tt [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+          /&nbsp [% nutrient.maximum %][% lang("points") %] ([% lang("nutriscore_source_value") %][% lang("sep") %]: [% nutrient.value %],
```
Comment (stephanegigandet) [inline templates/nutrition_score.tt:26]: the &nbsp in the code should be &nbsp; with a ;

File: templates/nutrition_score.tt (hunk 4 above)
Comment (stephanegigandet) [reply templates/nutrition_score.tt:25]: I just checked, we can do this instead:

[% lang("nutriscore_${group.type}_points") %]

[% lang("nutriscore_points_for_${nutrient.id}") %]

I think it's more clear than the [% lang("nutriscore_" _ group.type _ "_points") %] syntax.

**Comment by areeshatariq** [comment]
>     * is it possible to insert comments in the template that won't be rendered in the HTML: I think it's very important to let back-end devs document some things
Yes we can insert comments that won't be rendered in the HTML

>     * shouldn't we avoid conditionals IF-THEN-ELSE as much as we can: e.g.
> 
> 
> ```
>   [% IF is_beverage %]
>     <p> [% lang('nutriscore_is_beverage') %] </p>
>   [% ELSE %]
>     <p> [% lang('nutriscore_is_not_beverage') %] </p>
>   [% END %]
> ```
> 
> shouldn't we compute a `nutriscore_beverageness` in the back-end?

Yes, we did try to avoid as many conditions as we can to make the template easier to understand. I'll add this If Else to the back-end. Thank you

File: templates/nutrition_score.tt (hunk 5 above)
Comment (VaiTon) [inline templates/nutrition_score.tt:26]:
```suggestion
          /&nbsp; [% nutrient.maximum %][% lang("points") %] ([% lang("nutriscore_source_value") %][% lang("sep") %]: [% nutrient.value %], 
```

File: lib/ProductOpener/Display.pm [hunk 6]
```diff
@@ -8636,6 +8636,15 @@ sub display_nutriscore_calculation_details($) {
 
 	my $nutriscore_data_ref = shift;
 	
+	my $is_beverage;
```
Comment (VaiTon) [inline lib/ProductOpener/Display.pm:8639]: Would rename to "beverage_view" or something like this. The is_ prefix makes me think of a bool.

**Comment by stephanegigandet** [comment]
> * shouldn't we avoid conditionals IF-THEN-ELSE as much as we can: e.g.
> 
> ```
>   [% IF is_beverage %]
>     <p> [% lang('nutriscore_is_beverage') %] </p>
>   [% ELSE %]
>     <p> [% lang('nutriscore_is_not_beverage') %] </p>
>   [% END %]
> ```
> 
> shouldn't we compute a `nutriscore_beverageness` in the back-end?

Hi @CharlesNepote,

That's a very good question.




> * shouldn't we avoid conditionals IF-THEN-ELSE as much as we can: e.g.
> 
> ```
>   [% IF is_beverage %]
>     <p> [% lang('nutriscore_is_beverage') %] </p>
>   [% ELSE %]
>     <p> [% lang('nutriscore_is_not_beverage') %] </p>
>   [% END %]
> ```
> 
> shouldn't we compute a `nutriscore_beverageness` in the back-end?

That's a very good question @CharlesNepote .

@areeshatariq changed it to:

  <p> [% beverage_view %] </p>

It does make the template simpler, but we lost the connection to the string id in the .po file. That's another interesting question: should we try to put the calls to lang () to internationalize the strings in the template, or should we try to do all the calls to lang() in the Perl code.

The next statement is interesting too:
  
  [% IF is_fat %]
    <p> [% lang('nutriscore_proteins_is_added_fat') %] </p>
  [% END %]

We need the if statement because we don't want the <p> to be generated in the Perl code.

We could to the call to lang() in Perl though, and pass the translated message directly, in the "is_fat" variable (instead of just passing 0 or 1).

  [% IF is_fat %]
    <p> [% is_fat %] </p>
  [% END %]

**Comment by stephanegigandet** [comment]
> CONs:
> 
> * Github doesn't provide syntax highlighting for the template

@CharlesNepote , @areeshatariq : there is a simple solution to that: rename the .tt file in .html :) (it's in the /templates/ directory so it's obvious it's a template).

I just tried it, we have HTML highlighting in github: https://github.com/openfoodfacts/openfoodfacts-server/blob/areesha-stephane-wip/templates/nutrition_score.html

and also in the text / code editors:

image

**Comment by Ban3** [comment]
I added github syntax highlighting as HTML for the templates (per suggestion from @stephanegigandet). If there's a better option than HTML, see [here](https://github.com/github/linguist#overrides) how to configure it.

**Comment by stephanegigandet** [comment]
> I added github syntax highlighting as HTML for the templates (per suggestion from @stephanegigandet). If there's a better option than HTML, see [here](https://github.com/github/linguist#overrides) how to configure it.

Thanks @Ban3 . We probably will have templates for dynamically generated JS code as well (and possibly CSS). So I think the simplest thing is just to rename the template so that they have extension of what we are generating.  /templates/nutriscore.html (instead of .tt)

**Comment by hangy** [comment]
> We probably will have templates for dynamically generated JS code as well (and possibly CSS).

AFAIK, it is usually recommended to not create JS/CSS on-the-fly, but to use techniques like unobtrusive JS and to use selectors efficiently. 🙂 

> So I think the simplest thing is just to rename the template so that they have extension of what we are generating. /templates/nutriscore.html (instead of .tt)

The templates are not valid HTML/CSS/JS, so that working syntax highlighting basically depends on how forgiving the parser is. 🤷‍♂️

**Pull request merged**